File: remote/js/lib/RTRemoteServer.js
/**
 * Copyright (C) 2018 TopCoder Inc., All Rights Reserved.
 */

/**
 * The rt remote server class.
 *
 * @author      TCSCODER
 * @version     1.0
 */
const dgram = require('dgram');
const net = require('net');
const { URL } = require('url');
const RTEnvironment = require('./RTEnvironment');
const RTRemoteSerializer = require('./RTRemoteSerializer');
const RTRemoteTCPTransport = require('./RTRemoteTCPTransport');
const RTRemoteProtocol = require('./RTRemoteProtocol');
const RTRemoteMessageType = require('./RTRemoteMessageType');
const RTConst = require('./RTConst');
const RTMessageHelper = require('./RTMessageHelper');
const RTValueType = require('./RTValueType');
const RTException = require('./RTException');
const helper = require('./common/helper');
const logger = require('./common/logger');

/**
 * the registered object map
 * @type {object}
 */
const registeredObjectMap = {};

class RTRemoteServer {
  /**
   * create rtRemote server
   */
  constructor() {
    this.rpcSocketServer = null;
    this.rpcPort = null;
    this.rpcHost = null;
    this.serverName = `remote-server-${helper.getRandomUUID()}`;
  }

  /**
   * init rt remote server
   * 1. create udp multicast server, receive udp locate object packet and return message
   * 2. create rpc tcp socket server, receive rpc messages and return messages
   * @param {string} udpHost the multicast udp address
   * @param {number|int} udpPort the multicast udp port
   * @param {string} rpcHost the tcp bind ip address
   * @param {number|int} rpcPort the tcp bind port
   */
  init(udpHost, udpPort, rpcHost, rpcPort) {
    RTEnvironment.setRunMode(RTConst.SERVER_MODE);
    return this.openMulticastServer(udpHost, udpPort)
      .then(() => this.openRpcSocketServer(rpcHost, rpcPort))
      .then(() => this);
  }

  /**
   * open multicast server to receive udp locate object packet and return message
   * @param {string} udpHost the multicast udp address
   * @param {number|int} udpPort the multicast udp port
   * @return {Promise<dgram>} the promise with dgram
   */
  openMulticastServer(udpHost, udpPort) {
    const udpSocketIn = dgram.createSocket({
      type: 'udp4',
      reuseAddr: true,
    });

    const udpSocketOut = dgram.createSocket('udp4');
    udpSocketOut.bind(() => { // create channel to send located message
      udpSocketOut.setBroadcast(true);
      udpSocketOut.addMembership(udpHost);
    });

    return new Promise((resolve, reject) => {
      udpSocketIn.on('error', (err) => {
        logger.error(`multicast socket in error:\n${err.stack}`);
        reject(new RTException(err.message));
        udpSocketIn.close();
      });

      udpSocketIn.on('message', (buffer) => {
        const msg = RTRemoteSerializer.fromBuffer(buffer);
        const replyToURL = new URL(msg[RTConst.REPLY_TO]);
        const objectId = msg[RTConst.OBJECT_ID_KEY];

        if (registeredObjectMap[objectId]) {
          const locateMsg = RTMessageHelper.newLocateResponse(
            `tcp://${this.rpcHost}:${this.rpcPort}`,
            objectId, msg[RTConst.SENDER_ID], msg[RTConst.CORRELATION_KEY],
          );
          udpSocketOut.send(
            RTRemoteSerializer.toBuffer(locateMsg),
            replyToURL.port, replyToURL.hostname, (err) => {
              if (err) { // send failed, only log this error, no need do other actions
                logger.error(`send locate response to ${replyToURL.hostname}:${replyToURL.port} failed`);
                logger.error(err.stack);
              }
            },
          );
        } else {
          logger.warn(`client want search object ${objectId}, but not found in server ${this.serverName}.`);
        }
      });

      udpSocketIn.on('listening', () => {
        const address = udpSocketIn.address();
        logger.debug(`server bind multicast socket succeed, ${address.address}:${address.port}`);
        resolve(udpSocketIn);
      });
      udpSocketIn.bind({ port: udpPort }, () => {
        udpSocketIn.setBroadcast(true);
        udpSocketIn.addMembership(udpHost); // join to multicast channel
      });
    });
  }

  /**
   * open rpc tcp socket server receive rpc messages and return messages
   * @param {string} rpcHost the tcp bind ip address
   * @param {number|int} rpcPort the tcp bind port
   * @return {Promise<RTRemoteServer>} the remote server
   */
  openRpcSocketServer(rpcHost, rpcPort) {
    const that = this;
    return new Promise((resolve, reject) => {
      that.rpcSocketServer = net.createServer((socket) => {
        const transport = new RTRemoteTCPTransport(socket);

        // socket already created, no exceptions need catch for create protocol
        RTRemoteProtocol.create(transport, true).then((protocol) => {
          protocol.rtRemoteServer = that;
        });

        socket.on('error', (err) => {
          // only log error and close this socket, server still need continue run
          logger.error(err);
          socket.destroy();
        });
        socket.on('close', () => {
          logger.info('a client connection closed');
        });
      });

      this.rpcSocketServer.on('error', (err) => {
        logger.error(`unexpected rpc server error:\n${err.stack}`);
        reject(new RTException(err.message));
      });

      this.rpcSocketServer.listen(rpcPort, rpcHost, () => {
        that.rpcHost = that.rpcSocketServer.address().address;
        that.rpcPort = that.rpcSocketServer.address().port;
        logger.info(`rpc socket server bind succeed, ${that.rpcHost}:${that.rpcPort}`);
        resolve(that);
      });
    });
  }

  /**
   * get object by objectName/objectId
   *
   * @param {string} objName the object name
   */
  getObjectByName(objName) {
    const obj = registeredObjectMap[objName];
    if (!obj) {
      // here should not be throw null exception, it should return null, so that
      //    set property by name reponse can return OBJECT_NOT_FOUND
      logger.error(`getObjectByName object with name = ${objName} is null in server ${this.serverName}`);
    }
    return obj;
  }

  /**
   * hander client message
   * @param {RTRemoteTask} task the remote protocol
   * @return {Promise<any>} the promise with the result
   */
  handlerMessage(task) {
    switch (task.message[RTConst.MESSAGE_TYPE]) {
      case RTRemoteMessageType.GET_PROPERTY_BYNAME_REQUEST:
        return this.handlerGetPropertyByNameRequest(task);
      case RTRemoteMessageType.GET_PROPERTY_BYINDEX_REQUEST:
        return this.handlerGetPropertyByIndexRequest(task);
      case RTRemoteMessageType.SET_PROPERTY_BYNAME_REQUEST:
        return this.handlerSetPropertyByNameRequest(task);
      case RTRemoteMessageType.SET_PROPERTY_BYINDEX_REQUEST:
        return this.handlerSetPropertyByIndexRequest(task);
      case RTRemoteMessageType.METHOD_CALL_REQUEST:
        return this.handlerCallRequest(task);
      default:
        return Promise.reject(new RTException(`handlerMessage -> don't support message type ${
          task.message[RTConst.MESSAGE_TYPE]}`));
    }
  }

  /**
   * process set property by name request
   * @param {RTRemoteTask} task the remote task
   */
  handlerSetPropertyByNameRequest(task) {
    const { message } = task;
    if (message[RTConst.VALUE][RTConst.TYPE] === RTValueType.FUNCTION) {
      const rtFunction = message[RTConst.VALUE];
      message[RTConst.VALUE] = helper.updateListenerForRTFuction(task.protocol, rtFunction);
    }
    const response = helper.setProperty(this.getObjectByName(message[RTConst.OBJECT_ID_KEY]), message);
    return task.protocol.transport.send(RTRemoteSerializer.toBuffer(response));
  }

  /**
   * process set property by index request
   * @param {RTRemoteTask} task the remote task
   */
  handlerSetPropertyByIndexRequest(task) {
    const { message } = task;
    if (message[RTConst.VALUE][RTConst.TYPE] === RTValueType.FUNCTION) {
      const rtFunction = message[RTConst.VALUE];
      message[RTConst.VALUE] = helper.updateListenerForRTFuction(task.protocol, rtFunction);
    }
    const response = helper.setProperty(this.getObjectByName(message[RTConst.OBJECT_ID_KEY]), message);
    response[RTConst.MESSAGE_TYPE] = RTRemoteMessageType.SET_PROPERTY_BYINDEX_RESPONSE;
    return task.protocol.transport.send(RTRemoteSerializer.toBuffer(response));
  }

  /**
   * process set property by index request
   * @param {RTRemoteTask} task the remote task
   */
  handlerGetPropertyByIndexRequest(task) {
    const { message } = task;
    const response = helper.getProperty(this.getObjectByName(message[RTConst.OBJECT_ID_KEY]), message, this);
    response[RTConst.MESSAGE_TYPE] = RTRemoteMessageType.GET_PROPERTY_BYINDEX_RESPONSE;
    return task.protocol.transport.send(RTRemoteSerializer.toBuffer(response));
  }


  /**
   * process set property by name request
   * @param {RTRemoteTask} task the remote task
   */
  handlerGetPropertyByNameRequest(task) {
    const { message } = task;
    const response = helper.getProperty(this.getObjectByName(message[RTConst.OBJECT_ID_KEY]), message, this);
    return task.protocol.transport.send(RTRemoteSerializer.toBuffer(response));
  }

  /**
   * process call method request
   * @param {RTRemoteTask} task the remote task
   */
  handlerCallRequest(task) {
    const { message } = task;

    const args = message[RTConst.FUNCTION_ARGS] || [];
    if (args.length > 0) {
      args.forEach((arg, index) => {
        if (arg[RTConst.TYPE] === RTValueType.FUNCTION) {
          args[index] = helper.updateListenerForRTFuction(task.protocol, arg);
        }
      });
    }
    const response = helper.invokeMethod(this.getObjectByName(message[RTConst.OBJECT_ID_KEY]), message);
    return task.protocol.transport.send(RTRemoteSerializer.toBuffer(response));
  }

  /**
   * register a object with object name
   * @param {string} objectName the object name
   * @param {object} obj the object entity
   */
  registerObject(objectName, obj) {
    if (registeredObjectMap[objectName]) { // object exists
      throw new RTException(`object with name = ${objectName} already exists in server ${
        this.serverName}, please don't register again.`);
    }
    registeredObjectMap[objectName] = obj;
    logger.info(`object with name = ${objectName} register successfully in server ${this.serverName}`);
  }

  /**
   * check the objectName is register or not
   * @param objectName the object name
   */
  isRegister(objectName) { // eslint-disable-line
    return !!registeredObjectMap[objectName];
  }

  /**
   * unregister object from server
   *
   * @param {string} objectName the object name
   * @throws RTException throw errors if object not exist
   */
  unRegisterObject(objectName) {
    const object = registeredObjectMap[objectName];
    if (object == null) {
      throw new RTException(`cannot found register object named ${objectName} in server ${this.serverName}`);
    }
    registeredObjectMap[objectName] = null; // remove it
  }
}

/**
 * create rtRemote server with params
 * @param {string} udpHost the multicast udp address
 * @param {number|int} udpPort the multicast udp port
 * @param {string} rpcHost the tcp bind ip address
 * @param {undefined|number|int|null} rpcPort the tcp bind port, it will allocate by system if value is empty/null/undefined/0
 */
function create(udpHost, udpPort, rpcHost, rpcPort = null) {
  const server = new RTRemoteServer();
  return server.init(udpHost, udpPort, rpcHost, rpcPort || 0);
}

module.exports = {
  create,
};

File: remote/java/src/org/pxscene/rt/remote/RTRemoteServer.java
package org.pxscene.rt.remote;


import java.io.IOException;
import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.InetAddress;
import java.net.MulticastSocket;
import java.net.ServerSocket;
import java.net.Socket;
import java.net.URI;
import java.util.Map;
import java.util.Queue;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentLinkedDeque;
import org.apache.log4j.Logger;
import org.pxscene.rt.RTEnvironment;
import org.pxscene.rt.RTException;
import org.pxscene.rt.RTFunction;
import org.pxscene.rt.RTHelper;
import org.pxscene.rt.RTValue;
import org.pxscene.rt.RTValueType;
import org.pxscene.rt.remote.messages.RTMessageCallMethodRequest;
import org.pxscene.rt.remote.messages.RTMessageCallMethodResponse;
import org.pxscene.rt.remote.messages.RTMessageGetPropertyByNameRequest;
import org.pxscene.rt.remote.messages.RTMessageGetPropertyByNameResponse;
import org.pxscene.rt.remote.messages.RTMessageLocate;
import org.pxscene.rt.remote.messages.RTMessageSearch;
import org.pxscene.rt.remote.messages.RTMessageSetPropertyByNameRequest;
import org.pxscene.rt.remote.messages.RTMessageSetPropertyByNameResponse;

/**
 * the remote server
 */
public class RTRemoteServer {

  /**
   * the logger instance.
   */
  private static final Logger logger = Logger.getLogger(RTRemoteServer.class);

  /**
   * the register objects map, thread safe
   */
  private static final Map<String, Object> registerObjectMap = new ConcurrentHashMap<>();


  /**
   * the rt remote serializer instance
   */
  private RTRemoteSerializer serializer = new RTRemoteSerializer();

  /**
   * the task message queue
   */
  private Queue<RTRemoteTask> messageQueue;

  /**
   * the server socket instance
   */
  private ServerSocket serverSocket;

  /**
   * create new rtServer
   *
   * @param udpGroupAddr the multicast udp address
   * @param udpPort the multicast udp port
   * @param rpcAddr the tcp address
   * @param rpcPort the tcp port
   * @throws RTException raise error if create failed
   * @throws IOException rasie error if bind failed
   */
  public RTRemoteServer(InetAddress udpGroupAddr, int udpPort, InetAddress rpcAddr, int rpcPort)
      throws RTException, IOException {

    RTEnvironment.setRunMode(RTConst.SERVER_MODE); // set as server mode
    messageQueue = new ConcurrentLinkedDeque<>();
    MulticastSocket socketIn = new MulticastSocket(udpPort);
    DatagramSocket replySocketOut = new DatagramSocket();
    socketIn.joinGroup(udpGroupAddr);
    logger.debug("server bind multicast socket succeed");
    openRpcSocketServer(rpcAddr, rpcPort);
    new Thread(() -> { // start new thread to receive search object message
      while (true) {
        byte[] recvBuff = new byte[4096];
        DatagramPacket pkt = new DatagramPacket(recvBuff, recvBuff.length);
        try {
          socketIn.receive(pkt);
          RTMessageSearch search = serializer.fromBytes(recvBuff, 0, recvBuff.length);
          if (registerObjectMap.containsKey(search.getObjectId())) { // searched
            RTMessageLocate locate = new RTMessageLocate();
            locate.setEndpoint(new URI("tcp://" + rpcAddr.getHostAddress() + ":" + rpcPort));
            locate.setObjectId(search.getObjectId());
            locate.setSenderId(search.getSenderId());
            locate.setCorrelationKey(search.getCorrelationKey());
            byte[] locateBytes = serializer.toBytes(locate);
            replySocketOut.send(new DatagramPacket(locateBytes, 0, locateBytes.length,
                InetAddress.getByName(search.getReplyTo().getHost()),
                search.getReplyTo().getPort())); // send back locate message
          } else { // cannot found object, skip it
            logger.debug(
                "client want search object " + search.getObjectId() + ", but server not found");
          }
          Thread.sleep(1); // release cpu
        } catch (Exception e) {
          logger.error("udp read error", e);
          socketIn.close(); // close udp in socket
          replySocketOut.close(); // close udp out socket
          break;
        }
      }
    }).start();

    startProcessMessages();
  }

  /**
   * create create new rtServer
   *
   * @param udpGroupAddr the multicast udp address
   * @param udpPort the multicast udp port
   * @param rpcAddr the tcp address
   * @throws RTException raise error if create failed
   * @throws IOException rasie error if bind failed
   */
  public RTRemoteServer(InetAddress udpGroupAddr, int udpPort, InetAddress rpcAddr)
      throws RTException, IOException {
    this(udpGroupAddr, udpPort, rpcAddr, RTHelper.findRandomOpenPortOnAllLocalInterfaces());
  }

  /**
   * open socket server, so that can work as rpc server note that, serverSocket should not closed by
   * any exception, it should be work until closed by manually
   *
   * @param rpcAddr the tcp address
   * @param rpcPort the tcp port
   * @throws IOException if bind failed
   */
  private void openRpcSocketServer(InetAddress rpcAddr, int rpcPort) throws IOException {
    serverSocket = new ServerSocket(rpcPort, RTConst.BACK_LOG, rpcAddr);
    new Thread(() -> { // open new thread to accpet new connections
      while (true) {
        if (serverSocket == null || serverSocket.isClosed()) {
          break;
        }
        try {
          Socket socket = serverSocket.accept();
          RTRemoteTransport transport = new RTRemoteTCPTransport(socket);
          RTRemoteProtocol protocol = new RTRemoteProtocol(transport, true);
          protocol.setRtRemoteServer(this); // inject server to protocol
          Thread.sleep(1); // release cpu
        } catch (Exception e) {  // if accept failed, server should be log this error and wait next continue process next client
          logger.error("protocal error", e);
        }
      }
    }).start();
    logger.debug("rpc socket bind succeed.");
  }


  /**
   * start process message
   * create new thread to get message from task message queue
   */
  private void startProcessMessages() {
    new Thread(() -> { // create new thread to get message
      while (true) {
        try {
          if (messageQueue.size() > 0) {
            RTRemoteTask messageTask = messageQueue.poll(); // fetch a message
            handlerMessage(messageTask);
          }
          Thread.sleep(1); // release cpu
        } catch (Exception e) {
          logger.error("process message failed", e);
          break;
        }
      }
    }).start();
  }

  /**
   * get server object from cache
   *
   * @param name the object name
   * @return the server object
   */
  private Object getObjectByName(String name) {
    Object object = registerObjectMap.get(name);
    if (object == null) {
      throw new NullPointerException("object " + name + " should not be null");
    }
    return object;
  }

  /**
   * process message
   *
   * @param rtRemoteTask the task message
   * @throws RTException raise error if process failed
   */
  private void handlerMessage(RTRemoteTask rtRemoteTask) throws RTException {
    switch (rtRemoteTask.getMessage().getMessageType()) {
      case GET_PROPERTY_BYNAME_REQUEST:
        this.handlerGetPropertyByNameRequest(rtRemoteTask);
        break;
      case SET_PROPERTY_BYNAME_REQUEST:
        this.handlerSetPropertyByNameRequest(rtRemoteTask);
        break;
      case METHOD_CALL_REQUEST:
        this.handlerCallRequest(rtRemoteTask);
        break;
      default:
        throw new RTException(
            "handlerMessage -> don't support message type " + rtRemoteTask.getMessage()
                .getMessageType().toString() + "");
    }
  }

  /**
   * process call request message task
   *
   * @param rtRemoteTask the request message task
   * @throws RTException raise error if process failed
   */
  private void handlerCallRequest(RTRemoteTask rtRemoteTask) throws RTException {
    RTMessageCallMethodRequest callMethodRequest = (RTMessageCallMethodRequest) rtRemoteTask
        .getMessage();

    if (callMethodRequest.getFunctionArgs() != null
        && callMethodRequest.getFunctionArgs().size() > 0) {
      for (RTValue rtValue : callMethodRequest.getFunctionArgs()) {
        if (rtValue.getType().equals(RTValueType.FUNCTION)) {
          rtValue.setValue(RTHelper.updateListenerForRTFuction(rtRemoteTask.getProtocol(),
              (RTFunction) rtValue.getValue()));
        }
      }
    }
    RTMessageCallMethodResponse callMethodResponse = RTHelper
        .invokeMethod(getObjectByName(callMethodRequest.getObjectId()), callMethodRequest);
    rtRemoteTask.getProtocol().getTransport().send(serializer.toBytes(callMethodResponse));
  }


  /**
   * process get request message task
   *
   * @param rtRemoteTask the request message task
   * @throws RTException raise error if process failed
   */
  private void handlerGetPropertyByNameRequest(RTRemoteTask rtRemoteTask) throws RTException {
    RTRemoteMessage message = rtRemoteTask.getMessage();
    RTMessageGetPropertyByNameRequest getRequest = (RTMessageGetPropertyByNameRequest) message;
    RTMessageGetPropertyByNameResponse getResponse = RTHelper
        .getPropertyByName(getObjectByName(getRequest.getObjectId()), getRequest);
    rtRemoteTask.getProtocol().getTransport().send(serializer.toBytes(getResponse));
  }

  /**
   * process set request message task
   *
   * @param rtRemoteTask the request message task
   * @throws RTException raise error if process failed
   */
  private void handlerSetPropertyByNameRequest(RTRemoteTask rtRemoteTask) throws RTException {
    RTRemoteMessage message = rtRemoteTask.getMessage();
    RTMessageSetPropertyByNameRequest setRequest = (RTMessageSetPropertyByNameRequest) message;

    if (setRequest.getValue().getType().equals(RTValueType.FUNCTION)) { // update listener
      RTFunction oldFunction = (RTFunction) setRequest.getValue().getValue();
      setRequest.getValue()
          .setValue(RTHelper.updateListenerForRTFuction(rtRemoteTask.getProtocol(), oldFunction));
    }

    RTMessageSetPropertyByNameResponse setResponse = RTHelper
        .setPropertyByName(getObjectByName(setRequest.getObjectId()), setRequest);
    rtRemoteTask.getProtocol().getTransport().send(serializer.toBytes(setResponse));
  }


  /**
   * register a object with object name
   *
   * @param objectName the object name
   * @param object the object
   */
  public void registerObject(String objectName, Object object) {
    registerObjectMap.put(objectName, object);
    RTHelper.dumpObject(objectName, object);
  }


  /**
   * unregister object from server
   *
   * @param objectName the object name
   * @throws RTException throw errors if object not exist
   */
  public void unRegisterObject(String objectName) throws RTException {
    Object object = registerObjectMap.get(objectName);
    if (object == null) {
      throw new RTException("cannot found register object named " + objectName);
    }
    registerObjectMap.remove(objectName);
  }

  public Queue<RTRemoteTask> getMessageQueue() {
    return this.messageQueue;
  }
}
